### ISSUE TITLE
implement safe integer conversion

### ISSUE THREAD

**Issue by ftalbrecht** [wwu-numerik/dune-stuff#10]
See issue 279 on the user wiki: http://users.dune-project.org/issues/279

**Comment by ftalbrecht** [comment]
There is another thing I would like to discuss, which are integers as template arguments. Currently we use them mainly to indicate dimension, e.g. as in

    template< class DomainFieldType, int domainDim >
    class SomeFunction
    {
    public:
      static const unsigned int dimDomain = domainDim;
    };

This might be problematic and I would like to get some input on how to proceed.

* First of all I would like to replace the `static const unsigned int` by a `static const size_t` (I would leave the introduction of `constexpr` to a later point).
* The conversion in `dimDomain = domainDim` is technically unsafe and there are three possible ways to handle thin:
  * leave it as it is
  * add a `static_assert(domainDim >= 0, "Stupid?");` somewhere
  * change the template argument to `size_t`

Any input, @renemilk @sdrave @andreasbuhr?

This is of course related to #33.

**Comment by ftalbrecht** [comment]
Regarding the static case the following class:

```c++
template< size_t ii >
class Foo
{
public:
  Foo() { std::cout << ii << std::endl; }
};
```

can be instantiated with `Foo< 1 >();` while `Foo< -1 >();` gives a compiler error, at least with clang 3.5.1...

**Comment by ftalbrecht** [comment]
Implemented in pr #33.
